signature/jws/signer.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/notaryproject/notation-go-lib/crypto/jwsutil"
+	"github.com/notaryproject/notation-go-lib/crypto/timestamp"
+	"github.com/notaryproject/notation-go-lib/internal/crypto/pki"
+)
+
+// Signer signs artifacts and generates JWS signatures.
+type Signer struct {
+	// method is the method to sign artifacts.
+	method jwt.SigningMethod
+
+	// key is the signing key used to sign artifacts.
+	// The signing key can be either remote or local.
+	key crypto.PrivateKey
+
+	// keyID indicates which key is used to generate the signature.
+	keyID string
+
+	// certChain contains the X.509 public key certificate or certificate chain corresponding
+	// to the key used to generate the signature.
+	certChain [][]byte
+
+	// TSA is the TimeStamp Authority to timestamp the resulted signature if present.
+	TSA timestamp.Timestamper
+
+	// TSAVerifyOptions is the verify option to verify the fetched timestamp signature.
+	// The `Intermediates` in the verify options will be ignored and re-contrusted using
+	// the certificates in the fetched timestamp signature.
+	// An empty list of `KeyUsages` in the verify options implies ExtKeyUsageTimeStamping.
+	TSAVerifyOptions x509.VerifyOptions
+}
+
+// NewSigner creates a signer with the recommended signing method and a signing key bundled
+// with a certificate chain.
+// The relation of the provided siging key and its certificate chain is not verified,
+// and should be verified by the caller.
+func NewSigner(key crypto.PrivateKey, certChain []*x509.Certificate) (*Signer, error) {
+	method, err := SigningMethodFromKey(key)
+	if err != nil {
+		return nil, err
+	}
+	return NewSignerWithCertificateChain(method, key, certChain)
+}
+
+// NewSignerWithCertificateChain creates a signer with the specified signing method and
+// a signing key bundled with a certificate chain.
+// The relation of the provided siging key and its certificate chain is not verified,
+// and should be verified by the caller.
+func NewSignerWithCertificateChain(method jwt.SigningMethod, key crypto.PrivateKey, certChain []*x509.Certificate) (*Signer, error) {
+	if method == nil {
+		return nil, errors.New("nil signing method")
+	}
+	if key == nil {
+		return nil, errors.New("nil signing key")
+	}
+	if len(certChain) == 0 {
+		return nil, errors.New("missing signer certificate chain")
+	}
+
+	rawCerts := make([][]byte, 0, len(certChain))
+	for _, cert := range certChain {
+		rawCerts = append(rawCerts, cert.Raw)
+	}
+	return &Signer{
+		method:    method,
+		key:       key,
+		certChain: rawCerts,
+	}, nil
+}
+
+// NewSignerWithKeyID creates a signer with the specified signing method and a signing key
+// identified by a key ID.
+func NewSignerWithKeyID(method jwt.SigningMethod, key crypto.PrivateKey, keyID string) (*Signer, error) {
+	if key == nil {
+		return nil, errors.New("nil signing key")
+	}
+	if keyID == "" {
+		return nil, errors.New("empty signer key ID")
+	}
+
+	return &Signer{
+		method: method,
+		key:    key,
+		keyID:  keyID,
```

Comment:
Question: How is the KeyID used for signing and verification process since it's optional?

Reply:
As discussed in the recent meeting, the `KeyID` will be removed.